modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy

```diff
@@ -235,63 +235,74 @@ class TaskConfig extends LazyMap implements Cloneable {
         throw new IllegalArgumentException("Not a valid `ErrorStrategy` value: ${strategy}")
     }
 
+    def getResourceLimit(String directive) {
+        final limits = get('resourceLimits') as Map
+        return limits?.get(directive)
+    }
 
-    MemoryUnit getMemory() {
-        def value = get('memory')
-
-        if( !value )
-            return null
-
-        if( value instanceof MemoryUnit )
-            return (MemoryUnit)value
+    private MemoryUnit getMemoryUnit0(String key) {
+        getMemoryUnit0(get(key), key)
+    }
 
+    private MemoryUnit getMemoryUnit0(Object value, String key) {
         try {
-            new MemoryUnit(value.toString().trim())
+            return MemoryUnit.of(value)
         }
         catch( Exception e ) {
-            throw new AbortOperationException("Not a valid 'memory' value in process definition: $value")
+            throw new AbortOperationException("Not a valid `$key` value in process definition: $value")
         }
     }
 
+    MemoryUnit getMemory() {
+        def value = getMemoryUnit0('memory')
+        def limit = getMemoryUnit0(getResourceLimit('memory'), 'resourceLimits.memory')
+
+        if ( value && limit && value > limit )
+            value = limit
+
+        return value
+    }
+
     DiskResource getDiskResource() {
-        def value = get('disk')
+        def disk0 = get('disk')
+        if( !disk0 )
+            return null
 
-        if( value instanceof Map )
-            return new DiskResource((Map)value)
+        def disk = (disk0 instanceof Map) ? disk0 : [request: disk0]
 
-        if( value != null )
-            return new DiskResource(value)
+        def value = getMemoryUnit0(disk.request, 'disk')
+        def limit = getMemoryUnit0(getResourceLimit('disk'), 'resourceLimits.disk')
+        if ( value && limit && value > limit )
+            value = limit
 
-        return null
+        return new DiskResource(request: value, type: disk.type)
     }
 
     MemoryUnit getDisk() {
         getDiskResource()?.getRequest()
     }
 
     private Duration getDuration0(String key) {
-        def value = get(key)
-
-        if( !value )
-            return null
-
-        if( value instanceof Duration )
-            return (Duration)value
-
-        if( value instanceof Number )
-            return new Duration(value as long)
+        getDuration0(get(key), key)
+    }
 
+    private Duration getDuration0(Object value, String key) {
         try {
-            new Duration(value.toString().trim())
+            return Duration.of(value)
         }
         catch( Exception e ) {
             throw new AbortOperationException("Not a valid `$key` value in process definition: $value")
         }
-
     }
```

Comment:
What's the value of this? only throwing the AbortOperationException on error?

Reply:
I almost got rid of it, but `getDuration0` is useful to throw an error message with the correct name, as it could be `time`, `resourceLimits.time`, or `maxSubmitAwait`

Reply:
Ok, thanks. I've reshaped a bit to make it cleaner